### PR TITLE
Refactoring of P2P unit tests

### LINKED ISSUE #26: Use libp2p mocked network in tests
As [proposed](https://github.com/lazyledger/optimint/pull/17#discussion_r596850819) by @Wondertan it would be beneficial to use libp2p mocked network in tests. 

This require a change in `Client` to enable passing 'external' libp2p host. Right now host is created in `Start` method. New method `StartWithHost`?) may be introduced or functional options can be added to `Start`.

### PR DESCRIPTION
**Pull request by tzdybal**
Refactoring of P2P layer tests.

- [x] replace manual creation of clients with some "factory" function
- [x] use libp2p mock network

Resolves #26, #27.

### CODE REVIEW

File: p2p/utils_test.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	return
+}
+
+func (tn testNet) WaitForDHT() {
+	for i := range tn {
+		<-tn[i].dht.RefreshRoutingTable()
+	}
+}
+
+func makeTestNetowork(t *testing.T, n int, conns map[int][]int, logger log.Logger) testNet {
```
Comment (liamsi) [inline p2p/utils_test.go:40]: Yeah, this increases the readbility of the tests 👍🏼 

```suggestion
func makeTestNetwork(t *testing.T, n int, conns map[int][]int, logger log.Logger) testNet {
```
And the calling code as well.

File: p2p/utils_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	"github.com/lazyledger/optimint/config"
+	"github.com/lazyledger/optimint/log"
+)
+
+type testNet []*Client
+
+func (tn testNet) Start(ctx context.Context) (err error) {
+	for i := range tn {
+		err = multierr.Append(err, tn[i].Start(ctx))
```
Comment (liamsi) [inline p2p/utils_test.go:22]: This won't append anything if the `Start` doesn't return an error, right?

Comment (tzdybal) [reply p2p/utils_test.go]: yes, it only appends if `err != nil`.

File: p2p/utils_test.go (hunk 1 above)
Comment (tzdybal) [reply p2p/utils_test.go]: There should be :facepalm: reaction :D

File: p2p/utils_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	for i := range tn {
+		<-tn[i].dht.RefreshRoutingTable()
+	}
+}
+
+func startTestNetwork(t *testing.T, n int, conns map[int][]int, logger log.Logger) testNet {
+	t.Helper()
+	require := require.New(t)
+
+	mnet, err := mocknet.FullMeshLinked(context.Background(), n)
```
Comment (Wondertan) [inline p2p/utils_test.go:34]: Better create context at the beginning of the context and pass it here. That might be helpful in some tests

**Review by Wondertan — APPROVED** [review]
Looks good!

**Pull request merged**